gui/options.cpp

```diff
@@ -1885,52 +1880,73 @@ void GlobalOptionsDialog::apply() {
+	Common::String newCharset;
+	if (newLang != oldLang) {
+		TransMan.setLanguage(newLang);
+		ConfMan.set("gui_language", newLang);
+		newCharset = TransMan.getCurrentCharset();
+		isRebuildNeeded = true;
+	}
 #endif
-		Common::String oldTheme = g_gui.theme()->getThemeId();
-		if (g_gui.loadNewTheme(_newTheme)) {
+
+	GUI::ThemeEngine::GraphicsMode gfxMode = (GUI::ThemeEngine::GraphicsMode)_rendererPopUp->getSelectedTag();
+	Common::String oldGfxConfig = ConfMan.get("gui_renderer");
+	Common::String newGfxConfig = GUI::ThemeEngine::findModeConfigName(gfxMode);
+	if (newGfxConfig != oldGfxConfig) {
+		ConfMan.set("gui_renderer", newGfxConfig, _domain);
+	}
+
+	if (_newTheme.empty())
+		_newTheme = oldThemeId;
+
+	if (!g_gui.loadNewTheme(_newTheme, gfxMode, true)) {
+		Common::String errorMessage;
+
+		_curTheme->setLabel(oldThemeName);
+		_newTheme = oldThemeId;
+		ConfMan.set("gui_theme", _newTheme);
+		gfxMode = GUI::ThemeEngine::findMode(oldGfxConfig);
+		_rendererPopUp->setSelectedTag(gfxMode);
+		newGfxConfig = oldGfxConfig;
+		ConfMan.set("gui_renderer", newGfxConfig, _domain);
 #ifdef USE_TRANSLATION
-			// If the charset has changed, it means the font were not found for the
-			// new theme. Since for the moment we do not support change of translation
-			// language without restarting, we let the user know about this.
-			if (lang != TransMan.getCurrentLanguage()) {
-				TransMan.setLanguage(lang.c_str());
-				g_gui.loadNewTheme(oldTheme);
-				_curTheme->setLabel(g_gui.theme()->getThemeName());
-				MessageDialog error(_("The theme you selected does not support your current language. If you want to use this theme you need to switch to another language first."));
-				error.runModal();
-			} else {
+		if (newCharset != TransMan.getCurrentCharset()) {
+			errorMessage = _("Theme does not support selected language!");
+		} else {
 #endif
-				ConfMan.set("gui_theme", _newTheme);
+			errorMessage = _("Theme cannot be loaded!");
 #ifdef USE_TRANSLATION
-			}
-#endif
 		}
-		draw();
-		_newTheme.clear();
+		TransMan.setLanguage(oldLang);
+		_guiLanguagePopUp->setSelectedTag(selectedLang);
+		ConfMan.set("gui_language", oldLang);
+#endif
+		errorMessage += _("\nMisc settings will be restored.");
+		MessageDialog error(errorMessage);
+		error.runModal();
 	}
-#ifdef USE_TRANSLATION
-	Common::String oldLang = ConfMan.get("gui_language");
-	int selLang = _guiLanguagePopUp->getSelectedTag();
 
-	ConfMan.set("gui_language", TransMan.getLangById(selLang));
-
-	Common::String newLang = ConfMan.get("gui_language").c_str();
-	if (newLang != oldLang) {
-		// Activate the selected language
-		TransMan.setLanguage(selLang);
-
-		// Rebuild the Launcher and Options dialogs
-		g_gui.loadNewTheme(g_gui.theme()->getThemeId(), ThemeEngine::kGfxDisabled, true);
+	if (isRebuildNeeded) {
 		rebuild();
 		if (_launcher != 0)
 			_launcher->rebuild();
 	}
-#endif // USE_TRANSLATION
+	g_gui.loadNewTheme(_newTheme, gfxMode, true);
```

Comment:
I understand that if the first loadNewTheme() fails you need to call it again to set back the old theme. But can you explain why you need to call it again if it succeeds. You seem to imply in your comment that it needs to be called after the rebuild() (or that rebuild() needs to be called before loadNewTheme()). But as far as I know that is not the case (and the current code doesn't do it - it set the language, then calls loadNewTheme() then calls rebuild() and doesn't call again loadNewTheme()). We need to set the language before loadNewTheme() but this is already done before the first one.